lib/reports/overlap_report.rb

```diff
@@ -0,0 +1,149 @@
+      @ph = ph
+      @htdl = htdl
+      @sp = sp
+      validate!
+    end
+
+    # Check inputs
+    def validate!
+      valid = true
+      unless @organization.is_a?(String)
+        Services.logger.warn "organization must be a string"
+        valid = false
+      end
+
+      # Check boolish vals for boolishness.
+      boolish = {
+        ph: @ph,
+        htdl: @htdl,
+        sp: @sp
+      }
+      boolish.each do |var, val|
+        unless val.is_a?(TrueClass) || val.is_a?(FalseClass)
+          Services.logger.warn "#{var} must be a true/false"
+          valid = false
+        end
+      end
+      # But all boolish vals must not be false
+      if valid
+        if [@ph, @htdl, @sp].uniq == [false]
+          Services.logger.warn "At least one of ph/htdl/sp must be true"
+          valid = false
+        end
+      end
+
+      unless valid
+        raise "Check warnings, fix inputs, and run again."
+      end
+    end
+
+    # Fetch data, format, and output.
+    def run
+      File.open(outf_path, "w") do |outf|
+        outf.puts header.join("\t")
+        clusters.each do |cluster|
+          cluster_ph = count_cluster_ph(cluster)
+          cluster_htdl = count_cluster_htdl(cluster)
+          cluster_sp = count_cluster_sp(cluster)
+          cluster.holdings.select { |h| h.organization == @organization }.each do |hol|
+            outf.puts row(hol, cluster_ph, cluster_htdl, cluster_sp).join("\t")
+          end
+        end
+      end
+    end
+
+    # Get output location
+    def outf_path
+      if @outf_path.nil?
+        dir = Settings.overlap_report_path
+        if dir.nil?
+          Services.logger.warn "Settings.overlap_report_path not set, defaulting to /tmp"
+          dir = "/tmp"
+        end
+        ts = Time.now.strftime("%Y%m%d")
+        uuid = SecureRandom.uuid
+        @outf_path = "#{dir}/overlap_#{@organization}_#{ts}_#{uuid}.tsv"
+      end
+      @outf_path
+    end
+
+    # Array of header labels for the header line of the report
+    def header
+      @header ||= [
+        "ocn",
+        "local_id",
+        (@ph ? "ph_overlap" : nil),
+        (@htdl ? "htdl_overlap" : nil),
+        (@sp ? "sp_overlap" : nil)
+      ].compact # removes the nils
+    end
+
+    # Override method to get a more specific query, or pass in different param
+    def clusters(query: {"holdings.organization": @organization})
+      return enum_for(:clusters) unless block_given?
+
+      Cluster.where(query).each do |cluster|
+        yield cluster
+      end
+    end
+
+    # Formats an array of output values for a row of the report
+    def row(hol, cluster_ph, cluster_htdl, cluster_sp)
+      output = [
+        hol.ocn,
+        hol.local_id
+      ]
+      output << cluster_ph if @ph
+      output << cluster_htdl if @htdl
+      output << cluster_sp if @sp
+
+      output
+    end
+
+    # If @ph=true, count how many orgs have holdings in the cluster
+    def count_cluster_ph(cluster)
```

Comment:
I kind of wonder if each of these (ph, htdl, sp) should have their own little classes; then you instantiate whichever ones you want based on the provided options, and this becomes kind of a wrapper class that can include several of these kinds of counts? I'm not sure what exactly that might look like, just noticing the repetition several times in this class of the variation of ph/htdl/sp